Port FFmpeg 7.0 to ffmpegdirect Piers

basilgello opened the pull request:
From https://github.com/xbmc/xbmc/pull/24972

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello committed:
Port FFmpeg 7.1 to ffmpegdirect Piers

From https://github.com/xbmc/xbmc/pull/24972

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello commented:
@phunkyfish I tested that on Omega in Debian - works so far!

phunkyfish commented:
We can’t do this on omega, we don’t change ffmpeg for released versions. It needs to be on the Piers branch.

phunkyfish commented:
But should apply the same on Piers.

And nice work!

basilgello commented:
It id targeting Piers branch, despite the commit name!

phunkyfish commented:
Updated!

phunkyfish commented:
Can you rebase and update from the parent PR? That’s on 7.0.2 now and the Android build issues are fixed on Piers. So it should build fully on Jenkins.

Also, please check that any patches apply cleanly for ffmpeg. There are 5 altogether I think.

basilgello commented:
Roger that! :)

basilgello commented:
@phunkyfish the rebased original PR does not bring anything meaningful for FFmpegDirect:

```
git diff f89f51b8a67623f4882c455cd0f6615618096ebd..1d844951fd61d4a892cfbb5e22983a21c721bba7 -- $(git diff --name-only 031877fd4b833028b25a7d3f59f269907e9b4403..1d844951fd61d4a892cfbb5e22983a21c721bba7)
```

But I will update to 7.0.2 and ensure patches make sense

basilgello commented:
```patch
diff --git a/addons/resource.language.en_gb/resources/strings.po b/addons/resource.language.en_gb/resources/strings.po
index 0a67421cae..9ea4294d1f 100644
--- a/addons/resource.language.en_gb/resources/strings.po
+++ b/addons/resource.language.en_gb/resources/strings.po
@@ -11351,27 +11351,23 @@ msgstr ""
 
 #. Label for a select option or list item, representing an icon graphic (like a TV channel icon)
 #: xbmc/pvr/dialogs/GUIDialogPVRChannelManager.cpp
-#: xbmc/pvr/guilib/PVRGUIActionsEPG.cpp
 msgctxt "#19282"
 msgid "Current icon"
 msgstr ""
 
 #. Label for a select option or list item, representing an icon graphic (like a TV channel icon)
 #: xbmc/pvr/dialogs/GUIDialogPVRChannelManager.cpp
-#: xbmc/pvr/guilib/PVRGUIActionsEPG.cpp
 msgctxt "#19283"
 msgid "No icon"
 msgstr ""
 
-#. used by several skins and PVR
-#: xbmc/pvr/PVRContextMenus.cpp
+#. used by several skins
 msgctxt "#19284"
 msgid "Choose icon"
 msgstr ""
 
 #. Label for a select/menu option to select an icon graphic
 #: xbmc/pvr/dialogs/GUIDialogPVRChannelManager.cpp
-#: xbmc/pvr/guilib/PVRGUIActionsEPG.cpp
 msgctxt "#19285"
 msgid "Browse for icon"
 msgstr ""
@@ -17915,39 +17911,8 @@ msgctxt "#34113"
 msgid "To keep certain AVRs powered we send an inaudible random noise signal. You can disable this setting if you are using headphone or analog output."
 msgstr ""
 
-#. Indicates if Audio Engine should include lfe when downmixing
-#: system/settings/settings.xml
-msgctxt "#34114"
-msgid "Include LFE when downmixing"
-msgstr ""
-
-#. Description of setting with label #34114 "Include LFE when downmixing"
-#: system/settings/settings.xml
-msgctxt "#34115"
-msgid "If enabled, this setting will include lfe channel into the mixing when there is no dedicated LFE output channel available. This only makes sense for full range speakers."
-msgstr ""
-
-#. Description of setting with label #34114 "Include LFE when downmixing"
-#: system/settings/settings.xml
-msgctxt "#34116"
-msgid "Off"
-msgstr ""
-
-#. Description of setting with label #34114 "Include LFE when downmixing"
-#: system/settings/settings.xml
-msgctxt "#34117"
-msgid "50%"
-msgstr ""
-
-#. Description of setting with label #34114 "Include LFE when downmixing"
-#: system/settings/settings.xml
-msgctxt "#34118"
-msgid "100%"
-msgstr ""
-
-
 #empty strings from id 34114 to 34119
-#34119 reserved for future use
+#34114-34119 reserved for future use
 
 #: system/settings/settings.xml
 msgctxt "#34120"
diff --git a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEResampleFFMPEG.cpp b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEResampleFFMPEG.cpp
index a4fc91554f..6fce0af981 100644
--- a/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEResampleFFMPEG.cpp
+++ b/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEResampleFFMPEG.cpp
@@ -29,15 +29,8 @@ CActiveAEResampleFFMPEG::~CActiveAEResampleFFMPEG()
   swr_free(&m_pContext);
 }
 
-bool CActiveAEResampleFFMPEG::Init(SampleConfig dstConfig,
-                                   SampleConfig srcConfig,
-                                   bool upmix,
-                                   bool normalize,
-                                   double centerMix,
-                                   CAEChannelInfo* remapLayout,
-                                   AEQuality quality,
-                                   bool force_resample,
-                                   float sublevel)
+bool CActiveAEResampleFFMPEG::Init(SampleConfig dstConfig, SampleConfig srcConfig, bool upmix, bool normalize, double centerMix,
+                                   CAEChannelInfo *remapLayout, AEQuality quality, bool force_resample)
 {
   m_dst_chan_layout = dstConfig.channel_layout;
   m_dst_channels = dstConfig.channels;
@@ -141,9 +134,6 @@ bool CActiveAEResampleFFMPEG::Init(SampleConfig dstConfig,
     return false;
   }
 
-  if (sublevel > 0.0f)
-    av_opt_set_double(m_pContext, "lfe_mix_level", static_cast<double>(sublevel), 0);
-
   if (hasMatrix)
   {
     if (swr_set_matrix(m_pContext, reinterpret_cast<const double*>(m_rematrix), AE_CH_MAX) < 0)
```

basilgello commented:
Looks like 1h is definitely not enough for the pipeline.

phunkyfish commented:
> Looks like 1h is definitely not enough for the pipeline.

We in only need to worry about Jenkins. Everything else will timeout.

phunkyfish commented:
The patches do not apply cleanly. Can you fix them?

```
[ 87%] Performing patch step for 'ffmpeg'
patching file configure
Hunk #1 succeeded at 7060 (offset 332 lines).
patching file configure
Hunk #1 succeeded at 8188 (offset 363 lines).
patching file configure
Hunk #1 succeeded at 5784 (offset 218 lines).
Hunk #2 succeeded at 5837 (offset 220 lines).
patching file libavcodec/dxva2.c
Hunk #1 succeeded at 782 (offset 5 lines).
```

I don't see patch 4 being applied. Maybe it is no longer required.

basilgello commented:
I have removed one patch and renamed 0005 to 0004. OK I will re-export them

phunkyfish commented:
> I have removed one patch and renamed 0005 to 0004. OK I will re-export them

Perfect, thanks!

phunkyfish reviewed (approved):
Approved for when Kodi PR gets merged

zuzia-dev commented:
I tested under Debian 12 with KODI v22 Piers and it works great. Thanks!

basilgello commented:
@zuzia-dev You did build ffmpeg 7.0.2 for 12 or reverted the ffmpeg6 patch?

zuzia-dev commented:
> You did build ffmpeg 7.0.2 for 12 or reverted the ffmpeg6 patch?

@basilgello 
I compiled KODI v22 with ffmpeg 7.0.2: https://github.com/xbmc/xbmc/pull/24972

basilgello commented:
Thanks. In Debian 12 I will need to build against ffmpeg6, just in case :)

heitbaum commented:
successfully build tested against:
- https://github.com/xbmc/xbmc/pull/24972
- https://github.com/LibreELEC/LibreELEC.tv/pull/9569
- https://github.com/LibreELEC/LibreELEC.tv/pull/9570

heitbaum commented:
@basilgello @phunkyfish can this be merged and tagged for Piers?

basilgello commented:
Yes, there was an issue with ndk r27c which is partially related. Need to investigate this quickly

heitbaum commented:
> Yes, there was an issue with ndk r27c which is partially related. Need to investigate this quickly

Thanks for the update

phunkyfish commented:
iOS/tvos and Android all have build issues. We’ll need to fix those before merging.

basilgello committed:
Bump gnutls to 3.8.8

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello committed:
Fix build on darwin_embedded

VVC (H.266) implementation for AArch64 in 7.1 is incompatible with Apple CCAS.

Signed-off-by: Vasyl Gello <vasek.gello@gmail.com>

basilgello committed:
Release 22.1.0-Piers

phunkyfish commented:
@basilgello gnutls 3.8.9 is out now (since feb 8th), can you check to see if this version simplifies the patches for gnutls?
